util/Pos/Util/Verification.hs

```diff
@@ -0,0 +1,101 @@
+{- |
+
+Machinery to support pure datatype verification.
+
+1. Datatypes can have and not have internal pure consistency. If
+datatype doesn't have any implicit rules, it's called checked. 'Integer'
+is checked, 'Either Text String' is checked too. Otherwise, it's called
+unchecked. 'Coin' is unchecked, because it has implicit upper bound on the
+value, which is lower than 'Word64' bound. Newtype over string that is
+supposed to contain only strings of length less than 30 symbols is
+unchecked.
+
+2. Unchecked types' constructors should have prefix \"Unchecked\" (for
+example, 'UncheckedProxySecretKey'). They should also have description
+of why are they unchecked and provide some easy-to-use builders
+(@mkX@) if needed.
+
+3. If unchecked datatype @S@ is a part of @T@ (e.g. record field),
+then @T@ is not called unchecked. In other words, "being unchecked"
+property doesn't propagate, it is only related to the datatype itself.
+
+4. @T@ is called verifiable if it is unchecked or any of its parts are
+verifiable.
+
+5. Every verifiable type must have 'PVerifiable' instance. See
+the description of this class.
+
+-}
+module Pos.Util.Verification
+    (
+      PVer
+    , pverFail
+    , pverField
+
+    , PVerifiable(..)
+    , runPVerify
+    , runPVerifyText
+    , runPVerifyPanic
+    ) where
+
+import           Universum
+
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Text as T
+import           Formatting (sformat)
+import           Serokell.Util (VerificationRes (..), listJson, verResToMonadError)
+
+-- Prefix P is annoying, but we need to distinguish between pure
+-- verification and VAR somehow.
+
+-- | Pure verification monad. It accumulates errors along the
+-- verification (AST) path.
+newtype PVer a = PVer
+    { getPVer :: Either (NonEmpty Text) a
```

Comment:
Btw, there is also [Validation](https://www.stackage.org/haddock/nightly-2018-03-07/validation-1/Data-Validation.html).